peps/pep-0742.rst

```diff
@@ -0,0 +1,305 @@
+------------------------------------
+
+:pep:`724` previously proposed changing the specified behavior of :py:data:`typing.TypeGuard` so
+that if the return type of the guard is consistent with the input type, the behavior proposed
+here for ``TypeNarrower`` would apply. This proposal has some important advantages: because it
+does not require any runtime changes, it requires changes only in type checkers, making it easier
+for users to take advantage of the new, usually more intuitive behavior.
+
+However, this approach has some major problems. Users who have written ``TypeGuard`` functions
+expecting the existing semantics specified in :pep:`647` would see subtle and potentially breaking
+changes in how type checkers interpret their code. The split behavior of ``TypeGuard``, where it
+works one way if the return type is consistent with the input type and another way if it is not,
+could be confusing for users. The Typing Council was unable to come to an agreement in favor of
+:pep:`724`; as a result, we are proposing this alternative PEP.
+
+Do nothing
+----------
+
+Both this PEP and the alternative proposed in :pep:`724` have shortcomings. The latter are
+discussed above. As for this PEP, it introduces two special forms with very similar semantics,
+and it potentially creates a long migration path for users currently using ``TypeGuard``
+who would be better off with different narrowing semantics.
+
+One way forward, then, is to do nothing and live with the current limitations of the type system.
+However, we believe that the limitations of the current ``TypeGuard``, as outlined in the "Motivation"
+section, are significant enough that it is worthwhile to change the type system to address them.
+If we do not make any change, users will continue to encounter the same unintuitive behaviors from
+``TypeGuard``, and the type system will be unable to properly represent common type narrowing functions
+like ``inspect.isawaitable``.
+
+Open Issues
+===========
+
+Naming
+------
+
+This PEP currently proposes the name ``TypeNarrower``, emphasizing that the special form narrows
+the type of its argument. However, other names have been suggested, and we are open to using a
+different name.
+
+Options include:
+
+* ``IsInstance`` (`post by Paul Moore <https://discuss.python.org/t/pep-724-stricter-type-guards/34124/60>`__):
+  emphasizes that the new construct behaves similarly to the builtin :py:func:`isinstance`.
+* ``Predicate`` or ``TypePredicate``: mirrors TypeScript's name for the feature, "type predicates".
+* ``StrictTypeGuard`` (earlier drafts of :pep:`724`): emphasizes that the new construct performs a stricter
+  version of type narrowing than :py:data:`typing.TypeGuard`.
+* ``TypeCheck`` (`post by Nicolas Tessore <https://discuss.python.org/t/pep-724-stricter-type-guards/34124/59>`__):
+  emphasizes the binary nature of the check.
+* ``TypeIs``: emphasizes that the function returns whether the argument is of that type; mirrors
```

Comment:
FWIW, I really like the name `TypeIs`. I think it reads naturally in the code, communicates the two-way nature of the check (true means it is, false means it is not), and is concise.

In comparison, I find "TypeNarrower" to be unnecessarily long and somewhat grammatically clumsy.

I'm also not sure how well the term "narrow" (as it relates to types) communicates to the typical static typing user.

I also think there is benefit to having the name `TypeIs` be shorter and simpler than `TypeGuard` -- it implies (correctly) that the former is more commonly useful and less niche.

Reply:
Yes, I think it's a good candidate. I will mention this as a leading alternative when I post the PEP, and see if people like it.

We could go even shorter with simply `Is`: `def some_func(o: obj) -> Is[int]:`. But that might be a bit too short, and becomes hard to Google.

Reply:
> I also think there is benefit to having the name `TypeIs` be shorter and simpler than `TypeGuard` -- it implies (correctly) that the former is more commonly useful and less niche.

Agreed -- I find this an important criterion for choosing the name, and `TypeNarrower` fails. I find `TypeIs` slightly awkward (like all classes whose name is a verb) and would prefer `TypeCheck`.

Reply:
I don't like TypeCheck because it feels like that term is already in use by the term "type checker", and we say that a program "typechecks". But I could live with it if that's the consensus.

Reply:
I find `TypeNarrower` really cumbersome too.

Other ideas to consider: `Narrowed` (which is not unlike `Annotated`) or `NarrowedTo`.

Reply:
Added those to the list.